### ISSUE TITLE
Incorrect import path in generated ngfactory of an external library

### ISSUE THREAD

**Issue by igmdvc** [angular/angular#20006]
## I'm submitting a...

<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[X] Bug report  
[ ] Feature request
[ ] Documentation issue or request
[ ] Support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
</code></pre>

## Current behavior
generated component "ngfactory" js inside aot dir has the following import
`import * as i2 from "./FormComponent";`

## Expected behavior
generated component "ngfactory" js inside aot dir has the following import
`import { FormComponent } as i2 from "mymodule";`

## Minimal reproduction of the problem with instructions
my angular library tsconfig.json:
```
{
    "compilerOptions": {
        "module": "es2015",
        "target": "es6",
        "outDir": "dist",
        "rootDir": ".",
        "moduleResolution": "node",
        "noImplicitAny": true,
        "skipLibCheck": true,
        "allowSyntheticDefaultImports": true,
        "suppressImplicitAnyIndexErrors": true,
        "strictNullChecks": true,
        "experimentalDecorators": true,
        "emitDecoratorMetadata": true,
        "sourceMap": true,
        "removeComments": false,
        "declaration": true
    },
    "exclude": [
        "node_modules"
    ],
    "include": [
        "src/**/*.ts"
    ],
    "angularCompilerOptions": {
        "skipTemplateCodegen": true,
        "annotateForClosureCompiler": true,
        "annotationsAs": "static fields"
    }
}
````
angular library partial package.json:
```
{
    "main": "dist/server.umd.js",
    "module": "dist/client.es2015.js",
    "types": "dist/src/server.d.ts",
}
```
application consuming library tsconfig.json:
```
{
    "compilerOptions": {
        "target": "es5",
        "module": "es2015",
        "moduleResolution": "node",
        "noImplicitAny": true,
        "skipLibCheck": true,
        "allowSyntheticDefaultImports": true,
        "suppressImplicitAnyIndexErrors": true,
        "strictNullChecks": true,
        "experimentalDecorators": true,
        "emitDecoratorMetadata": true,
        "sourceMap": true,
        "removeComments": false,
        "lib": ["es2015", "dom"],
        "outDir": "compiled/aot",
        "baseUrl": ".",
        "paths": {
            "@angular/*": ["node_modules/@angular/*"]
        }
    },
    "files": [
        "compiled/src/frontend-application/AppModule",
        "compiled/src/frontend-application/bootstrap"
    ],
    "angularCompilerOptions": {
        "skipTemplateCodegen": false,
        "skipMetadataEmit": true,
        "annotationsAs": "static fields",
        "annotateForClosureCompiler": true
    }
}
```

## What is the motivation / use case for changing the behavior?
Generated ngfactory includes imports with relative paths for external modules, which is wrong, as external modules should have an unique entry point based on package.json module entry

## Environment

<pre><code>
Angular version: 5.0.0.rc.8

Browser:
- [ ] Chrome (desktop) version XX
- [ ] Chrome (Android) version XX
- [ ] Chrome (iOS) version XX
- [ ] Firefox version XX
- [ ] Safari (desktop) version XX
- [ ] Safari (iOS) version XX
- [ ] IE version XX
- [ ] Edge version XX
 
For Tooling issues:
- Node version: XX  
- Platform:  

Others:

</code></pre>

**Comment by sebek64** [comment]
I'm seeing the same problem with rc9. In my case, it is for ngx-bootstrap library. With angular 4, the generated import is in the form

import * as i2 from 'ngx-bootstrap/alert/alert.component';

while now, it is

import * as i2 from "./alert.component";

This breaks subsequent rollup bundling.

**Comment by bgarel** [comment]
I have the same problem with all my external libraries having a .ngfactory.js generated by ngc.
As a workaround, after the ngc process, I copy all the required node_modules directories to the aot dir.
example :
`cp -r node_modules/@ng-bootstrap/ng-bootstrap/  xxxxxxx/aot/node_modules/@ng-bootstrap/`

And then, I can run webpack

**Comment by igmdvc** [comment]
@bgarel yes that is what I thought at first. What I ended up doing is not having outDir (I copy source files to a different source). ngfactory files from node_modules are generated in the corresponding node_modules folder and paths are right so I can bundle everything.

Anyway, I think this issue should be prioritized properly or at least provide a non hacky solution. What annoys me is that other third parties like material2 ngfactory files have correct path

**Comment by jasonaden** [comment]
@igmdvc Can you create a reproduction for this issue? There's not enough info here to debug. I'm not sure if you're pre-creating the ngfactory files on the library side (you shouldn't be) or if you're using a rolled up index file (if you are, you're missing angularCompilerOptions settings).

**Comment by bgarel** [comment]
I have created a github repo to reproduce this issue.
[angular-bug-20006](https://github.com/bgarel/angular-bug-20006)

**Comment by mlc-mlapis** [comment]
@jasonaden ... here is is the basic reproduction of the problem ... when using external packages `primeng` and `angular-modal-gallery`. You can see it here https://github.com/mlc-mlapis/demo-angular-modal-gallery

Reproduction steps:
* clone the repo
* npm install
* npm run ngc
* gulp serve:aot

You can see the following imports in `app.component.ngfactory.js` file:
```
var i2 = require("../node_modules/primeng/components/slider/slider.ngfactory");
var i5 = require("../node_modules/angular-modal-gallery/angular-modal-gallery.ngfactory");
```
when they should be:
```
var i2 = require("primeng/components/slider/slider.ngfactory");
var i5 = require("angular-modal-gallery/angular-modal-gallery.ngfactory");
```

**Comment by theodorejb** [comment]
I'm having this exact same issue when trying to AOT-compile an app that depends on ag-Grid.

**Comment by Metatagross** [comment]
I have the same problem with primeng. They said that clearing include tag helps but in my case i don't even have that tag.

**Comment by alxhub** [comment]
Looking through the repro, I can't figure out if this is WAI or an actual bug as I don't fully understand the path mechanism within the generated `ngfactory` files.

That said, the good news is that Ivy will completely solve this problem!

**Comment by alxhub** [comment]
Closing as this is resolved in 9.0 via the Ivy compiler!
